src/main/java/fi/aalto/cs/apluscourses/intellij/model/CourseUpdater.java

```diff
@@ -0,0 +1,147 @@
+                       @NotNull URL courseUrl,
+                       @NotNull Event eventToTrigger,
+                       @NotNull Notifier notifier,
+                       long updateInterval) {
+    this.course = course;
+    this.project = project;
+    this.courseUrl = courseUrl;
+    this.eventToTrigger = eventToTrigger;
+    this.notifier = notifier;
+    this.updateInterval = updateInterval;
+  }
+
+  /**
+   * Starts or restarts this updater. This method is thread safe and can be called from multiple
+   * threads.
+   */
+  public synchronized void restart() {
+    if (thread != null) {
+      thread.interrupt();
+    }
+    thread = new Thread(this::run);
+    thread.start();
+  }
+
+  /*
+   * The run lock ensures that only one thread executes the 'run' method. The run lock guarantees
+   * that a new thread created by 'restart' blocks until the previous thread has received the
+   * interrupt and exited. Note, that making the 'run' method 'synchronized' would cause a deadlock,
+   * because 'restart' is synchronized as well.
+   */
+  private final Object runLock = new Object();
+
+  private void run() {
+    synchronized (runLock) {
+      while (true) { //  NOSONAR
+        try {
+          var newCourse = fetchCourse();
+          if (newCourse != null) {
+            updateModuleIds(newCourse);
+            eventToTrigger.trigger();
+          }
+          var updatableModules = course.getUpdatableModules()
+              .stream()
+              .filter(m -> notifiedModules.add(m.getName()))
+              .collect(Collectors.toList());
+          if (!updatableModules.isEmpty()) {
+            moduleUpdatesNotification = new NewModulesVersionsNotification(updatableModules);
+            notifier.notifyAndHide(moduleUpdatesNotification, project);
+          }
+          Thread.sleep(updateInterval); // Good night :-)
+        } catch (InterruptedException e) {
+          Thread.currentThread().interrupt();
+          return;
+        }
+      }
+    }
+  }
+
+  @Nullable
+  private Course fetchCourse() {
+    try {
+      // Read-lock needed, since we access the project
+      return ReadAction.compute(() -> Course.fromUrl(courseUrl, new IntelliJModelFactory(project)));
+    } catch (Exception e) {
+      return null;
+    }
+  }
+
+  private void updateModuleIds(@NotNull Course newCourse) {
+    var urisToModules = createUriToModuleMapping(newCourse);
+    for (var module : course.getModules()) {
+      var newModule = urisToModules.get(urlToUri(module.getUrl()));
+      if (newModule != null) {
+        module.updateVersionId(newModule.getVersionId());
+      }
+    }
+  }
+
+  @Nullable
+  private static URI urlToUri(@NotNull URL url) {
+    try {
+      return url.toURI();
+    } catch (URISyntaxException e) {
+      return null;
+    }
+  }
+
+  private static Map<URI, Module> createUriToModuleMapping(@NotNull Course course) {
+    // hashCode and equals of URLs can cause DNS lookups, so URIs are preferred in maps
```

Comment:
That's... interesting. o.O Nice catch, anyways!

Reply:
This WTF-fact was actually noted by sonar :smile: One of the rare moments where I feel grateful for its existence